src/services/refactors/convertToNamedParameters.ts

```diff
@@ -0,0 +1,510 @@
+
+        function isValidVariableDeclaration(node: Node): node is ValidVariableDeclaration {
+            return isVariableDeclaration(node) && isVarConst(node) && isIdentifier(node.name) && !node.type;
+        }
+    }
+
+    function hasThisParameter(parameters: NodeArray<ParameterDeclaration>): boolean {
+        return parameters.length > 0 && isThis(parameters[0].name);
+    }
+
+    function getRefactorableParametersLength(parameters: NodeArray<ParameterDeclaration>): number {
+        if (hasThisParameter(parameters)) {
+            return parameters.length - 1;
+        }
+        return parameters.length;
+    }
+
+    function getRefactorableParameters(parameters: NodeArray<ValidParameterDeclaration>): NodeArray<ValidParameterDeclaration> {
+        if (hasThisParameter(parameters)) {
+            parameters = createNodeArray(parameters.slice(1), parameters.hasTrailingComma);
+        }
+        return parameters;
+    }
+
+    function createNewArgument(functionDeclaration: ValidFunctionDeclaration, functionArguments: NodeArray<Expression>): ObjectLiteralExpression {
+        const parameters = getRefactorableParameters(functionDeclaration.parameters);
+        const hasRestParameter = isRestParameter(last(parameters));
+        const nonRestArguments = hasRestParameter ? functionArguments.slice(0, parameters.length - 1) : functionArguments;
+        const properties = map(nonRestArguments, (arg, i) => {
+            const property = createPropertyAssignment(getParameterName(parameters[i]), arg);
+            suppressLeadingAndTrailingTrivia(property.initializer);
+            copyComments(arg, property);
+            return property;
+        });
+
+        if (hasRestParameter && functionArguments.length >= parameters.length) {
+            const restArguments = functionArguments.slice(parameters.length - 1);
+            const restProperty = createPropertyAssignment(getParameterName(last(parameters)), createArrayLiteral(restArguments));
+            properties.push(restProperty);
+        }
+
+        const objectLiteral = createObjectLiteral(properties, /*multiLine*/ false);
+        return objectLiteral;
+    }
+
+    function createNewParameters(functionDeclaration: ValidFunctionDeclaration, program: Program, host: LanguageServiceHost): NodeArray<ParameterDeclaration> {
+        const refactorableParameters = getRefactorableParameters(functionDeclaration.parameters);
+        const bindingElements = map(refactorableParameters, createBindingElementFromParameterDeclaration);
+        const objectParameterName = createObjectBindingPattern(bindingElements);
+        const objectParameterType = createParameterTypeNode(refactorableParameters);
+
+        let objectInitializer: Expression | undefined;
+        // If every parameter in the original function was optional, add an empty object initializer to the new object parameter
+        if (every(refactorableParameters, param => !!param.initializer || !!param.questionToken)) {

```

Comment:
I think there is an isOptional[Parameter] in utilities.ts or in checker.ts. Can you take a look to see whether you could use that if it exists?